### PR TITLE
Rama/Suhas | Hive-120097 | Showing warning while logging out with dra…

### PR DESCRIPTION
**Pull request by suhaskatkuri-wq**
Screenshot 2026-07-10 at 10 45 47 AM

Screenshot 2026-07-10 at 11 15 02 AM

Hive card : [Showing warning while logging out with drafts present](https://app.hive.com/workspace/adoBEQk5LKDDYFd8b?projectId=sfxBCpc3JKca3H2S8&actionViewId=aLgrTmXYc4HRqLbXw&tabId=thBjsTJW9AaHvQQBr&actionId=XzaRZCBvfGj9FMHvA)

### CODE REVIEW

File: ui/app/common/auth/authentication.js [hunk 1]
```diff
@@ -236,21 +236,73 @@ angular.module('authentication')
...
                 logoutUser();
+            };
+            draftsWarningScope.cancel = function () {
+                ngDialog.close(dialog.id);
+            };
+        }
+
+        function isLogoutDraftsWarningEnabled () {
+            return configurationService.getConfigurations(['enableLogoutDraftsWarning']).then(function (response) {
+                return response.enableLogoutDraftsWarning !== 'false';
```
Comment (SooryaKumaranC-tw) [inline ui/app/common/auth/authentication.js]: Have you checked the default flow, when the **enableLogoutDraftsWarning** is not given or null? 

What happens at that time?

File: ui/app/adt/controllers/careViewController.js [hunk 2]
```diff
@@ -21,12 +21,7 @@ angular.module('bahmni.adt')
             $state.go('home');
         },
         onLogOut: function () {
-            auditLogService.log(undefined, 'USER_LOGOUT_SUCCESS', undefined, 'MODULE_LABEL_LOGOUT_KEY').then(function () {
```
Comment (SooryaKumaranC-tw) [inline ui/app/adt/controllers/careViewController.js:24]: Why did you remove the audit log? This is needed right. I can't see in logoutService.attemptLogout you have logged the audit. Please confirm

File: ui/app/common/auth/init.js [hunk 3]
```diff
@@ -2,4 +2,4 @@
 var Bahmni = Bahmni || {};
 Bahmni.Auth = Bahmni.Auth || {};
 
-angular.module('authentication', ['ui.router']);
+angular.module('authentication', ['ui.router', 'bahmni.common.services', 'bahmni.common.domain', 'ngDialog']);
```
Comment (SooryaKumaranC-tw) [inline ui/app/common/auth/init.js]: Do we need both the module? And can you explain for what they are used

File: ui/app/common/i18n/services/mergeLocaleFilesService.js [hunk 4]
```diff
@@ -29,15 +37,17 @@ angular.module('bahmni.common.i18n')
                 };
 
                 var mergeTranslations = function (result) {
-                    var baseFileData = result[0] ? result[0].data : undefined;
-                    var customFileData = result[1] ? result[1].data : undefined;
+                    var commonFileData = result[0] ? result[0].data : undefined;
+                    var appFileData = result[1] ? result[1].data : undefined;
+                    var customFileData = result[2] ? result[2].data : undefined;
```
Comment (SooryaKumaranC-tw) [inline ui/app/common/i18n/services/mergeLocaleFilesService.js:42]: Why are we introducing this new result[2] ? result[2].data ?

File: ui/app/common/domain/services/configurationService.js [hunk 5]
```diff
@@ -219,6 +219,18 @@ angular.module('bahmni.common.domain')
             });
         };
 
+        configurationFunctions.enableLogoutDraftsWarning = function () {
+            return $http.get(Bahmni.Common.Constants.globalPropertyUrl, {
+                params: {
+                    property: 'bahmni.enableLogoutDraftsWarning'
```
Comment (SooryaKumaranC-tw) [inline ui/app/common/domain/services/configurationService.js]: Why did we choose 'bahmni.enableLogoutDraftsWarning'  ? Is there any reason for this path? Is there any other apt path for this **enableLogoutDraftsWarning** flow ?

File: ui/app/orders/index.html [hunk 6]
```diff
@@ -5,10 +5,10 @@
         <link rel="shortcut icon" type="image/x-icon" href="../favicon.ico">
         <link rel="stylesheet" href="../lib/jquery/jquery-ui-1.10.4.custom.min.css">
         <!-- build:css orders.min.css -->
-        <link rel="stylesheet" href="../styles/clinical.css"/>
         <link rel="stylesheet" href="../components/ngDialog/css/ngDialog.min.css"/>
         <link rel="stylesheet" href="../components/ngDialog/css/ngDialog-theme-default.min.css"/>
         <link rel="stylesheet" href="../components/ngDialog/css/ngDialog-theme-plain.min.css"/>
+        <link rel="stylesheet" href="../styles/clinical.css"/>
```
Comment (SooryaKumaranC-tw) [inline ui/app/orders/index.html]: I see this line was already there. So you can remove this change

File: ui/app/registration/index.html [hunk 7]
```diff
@@ -7,12 +7,12 @@
         <link rel="shortcut icon" type="image/x-icon" href="../favicon.ico">
         <link rel="stylesheet" href="../lib/jquery/jquery-ui-1.10.4.custom.min.css">
         <link rel="stylesheet" href="../components/select2/select2.css"/>
-        <!-- build:css registration.min.css -->
-        <link rel="stylesheet" href="../styles/registration.css"/>
-        <!-- endbuild -->
         <link rel="stylesheet" href="../components/ngDialog/css/ngDialog.min.css"/>
         <link rel="stylesheet" href="../components/ngDialog/css/ngDialog-theme-default.min.css"/>
         <link rel="stylesheet" href="../components/ngDialog/css/ngDialog-theme-plain.min.css"/>
+        <!-- build:css registration.min.css -->
```
Comment (SooryaKumaranC-tw) [inline ui/app/registration/index.html:10]: This changes also already exist. You can remove this file

File: ui/test/unit/adt/controllers/careViewController.spec.js [hunk 8]
```diff
@@ -10,21 +10,10 @@ describe("CareViewController", function () {
...
     auditLogService = jasmine.createSpyObj('auditLogService', ['log']);
-    sessionService = jasmine.createSpyObj('sessionService', ['destroy']);
+    logoutService = jasmine.createSpyObj('logoutService', ['attemptLogout']);
     auditLogService.log.and.returnValue({
         then: function(callback) { return callback(); }
     });
-    sessionService.destroy.and.returnValue({
-        then: function() { }
-    });
-    auditLogService = jasmine.createSpyObj('auditLogService', ['log']);
```
Comment (SooryaKumaranC-tw) [inline ui/test/unit/adt/controllers/careViewController.spec.js:20]: Do add this  ->  auditLogService = jasmine.createSpyObj('auditLogService', ['log']);

We need to log the logout as before. Do not change the existing behaviour

File: ui/app/common/auth/authentication.js (hunk 1 above)
Comment (suhaskatkuri-wq) [reply ui/app/common/auth/authentication.js]: We are defaulting "enableLogoutDraftsWarning" - when it's not given or null.

File: ui/app/adt/controllers/careViewController.js (hunk 2 above)
Comment (suhaskatkuri-wq) [reply ui/app/adt/controllers/careViewController.js:24]: We haven't removed it, instead we moved it to, so that logoutService.logoutUser()  still fires the same  auditLogService.log(undefined, 'USER_LOGOUT_SUCCESS', ...)  call, and now it runs on every logout path instead of being duplicated in each controller.

File: ui/app/common/auth/init.js (hunk 3 above)
Comment (suhaskatkuri-wq) [reply ui/app/common/auth/init.js]: Yes, both are required —  bahmni.common.services  for  formDraftService  (pre-existing, for draft detection) and  bahmni.common.domain  for  configurationService  (new, for the GP toggle).

File: ui/app/common/i18n/services/mergeLocaleFilesService.js (hunk 4 above)
Comment (suhaskatkuri-wq) [reply ui/app/common/i18n/services/mergeLocaleFilesService.js:42]: We are introducing this because It's related to the drafts-warning modal strings so they don't need to be duplicated in all 11 apps' locale files (added a new shared  i18n/common/locale_en.json).  mergeLocaleFilesService  now merges 3 sources instead of 2.

File: ui/app/common/domain/services/configurationService.js (hunk 5 above)
Comment (suhaskatkuri-wq) [reply ui/app/common/domain/services/configurationService.js]: This was suggested by the agent, as the naming convention was used by similar logout-related properties in the same file ( bahmni.quickLogoutComboKey ,  bahmni.contextCookieExpirationTimeInMinutes )

File: ui/app/orders/index.html (hunk 6 above)
Comment (suhaskatkuri-wq) [reply ui/app/orders/index.html]: Yeah, Sorry, this diff is unneccesary, removed it.
Thank you

File: ui/test/unit/adt/controllers/careViewController.spec.js (hunk 8 above)
Comment (suhaskatkuri-wq) [reply ui/test/unit/adt/controllers/careViewController.spec.js:20]: Moved this assertion to  logOutDirective.spec.js , which tests the actual  logoutService.logoutUser()  behavior — added  expect(auditLogService.log).toHaveBeenCalledWith(...)  there. Re-adding it in  careViewController.spec.js  isn't meaningful anymore since the controller no longer calls  auditLogService  directly, only  logoutService.attemptLogout

File: ui/app/registration/index.html (hunk 7 above)
Comment (suhaskatkuri-wq) [reply ui/app/registration/index.html:10]: Yeah, sorry again, same here, diff is unneccesary , so reverting
Thank you

**Review by SooryaKumaranC-tw — APPROVED** [review]

File: ui/app/common/auth/authentication.js [hunk 9]
```diff
@@ -236,21 +236,73 @@ angular.module('authentication')
...
+            }, function () {
+                return true;
+            });
+        }
+
+        function attemptLogout (scope) {
```
Comment (vvkpd) [inline ui/app/common/auth/authentication.js:262]: suggestion: No guard against concurrent `attemptLogout` calls. If the user double-clicks the logout button before the `hasDraftsForProvider` HTTP call resolves, `attemptLogout` runs twice — and if drafts exist, two warning dialogs open simultaneously.

Consider adding an in-flight guard:
```javascript
var isAttemptingLogout = false;
function attemptLogout (scope) {
    if (isAttemptingLogout) { return; }
    isAttemptingLogout = true;
    // reset flag in cancel callback and on direct logout
}
```

File: ui/app/common/auth/authentication.js [hunk 10]
```diff
@@ -236,21 +236,73 @@ angular.module('authentication')
...
+            if ((event.metaKey || event.ctrlKey) && event.key === $rootScope.quickLogoutComboKey) {
+                logoutService.attemptLogout(scope);
             }
         }
         return {
             link: function (scope, element) {
                 element.bind('click', function () {
                     scope.$apply(function () {
-                        logoutUser();
+                        logoutService.attemptLogout(scope);
```
Comment (vvkpd) [inline ui/app/common/auth/authentication.js]: nit: The click handler wraps `attemptLogout` in `scope.$apply()`, but the keyboard handler (line 289-291) calls it directly without `$apply`. Since `attemptLogout` uses `$http`/`$q` promises which already schedule digests, neither path strictly needs `$apply`. The inconsistency could cause a `$apply already in progress` error in edge cases.

Suggestion: remove `scope.$apply()` from the click handler.

File: ui/test/unit/common/auth/logOutDirective.spec.js [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+        element.triggerHandler('click');
+
+        setTimeout(function () {
+            expect(ngDialog.open).not.toHaveBeenCalled();
+            expect(sessionService.destroy).toHaveBeenCalled();
+            done();
+        }, 0);
+    });
+});
```
Comment (vvkpd) [inline ui/test/unit/common/auth/logOutDirective.spec.js:217]: suggestion: This spec covers the click path thoroughly (6 scenarios) but has zero coverage for the keyboard shortcut path (`Ctrl/Cmd + quickLogoutComboKey`). The keyboard handler follows a different code path (no `$apply` wrapper) so a regression there would go undetected.

Suggested additions:
1. Keyboard shortcut with no drafts → logs out directly
2. Keyboard shortcut with drafts → shows warning dialog
3. Keyboard shortcut with feature disabled → logs out directly

File: ui/app/i18n/common/locale_en.json [hunk 12]
```diff
@@ -0,0 +1,6 @@
+{
+    "LOGOUT_DRAFTS_WARNING_TITLE_KEY": "Draft in Progress",
+    "LOGOUT_DRAFTS_WARNING_MESSAGE_KEY": "You have unsaved observation forms. Are you sure you want to logout ?",
+    "LOGOUT_DRAFTS_WARNING_CANCEL_KEY": "Cancel",
+    "LOGOUT_DRAFTS_WARNING_LOGOUT_KEY": "Logout"
+}
```
Comment (vvkpd) [inline ui/app/i18n/common/locale_en.json:6]: nit: Only `locale_en.json` is added. Other apps (clinical, adt, etc.) have locale files for multiple languages. Users with a non-English locale will see raw translation keys in the dialog. Either add equivalent files for supported languages, or document this as a known limitation for this release.

**Review by vvkpd — COMMENTED** [review]
## Review for Hive-120097

Clean implementation of the drafts warning feature. The service extraction is well-done, the fail-safe design is correct, and Soorya's earlier concerns are all properly addressed. Found 3 things worth discussing before merge.

Found 0 blocking issues, 3 suggestions, and 1 nit.

**Acceptance Criteria**: 3/3 met (all AC and popup behaviors satisfied)

### Issues not in diff

| File:Line | Type | Issue |
|-----------|------|-------|
| ui/app/registration/controllers/navigationController.js:17 | suggestion | `$scope.logout` is dead code that bypasses `logoutService` — template uses the `log-out` directive so this is never called, but it's a maintenance hazard. Either delete the function or update it to call `logoutService.attemptLogout($scope)`. |

**Comment by vvkpd** [comment]
@PoojaSastry-TW please review these prs

File: ui/app/i18n/common/locale_en.json (hunk 12 above)
Comment (SooryaKumaranC-tw) [reply ui/app/i18n/common/locale_en.json:6]: +1

File: ui/app/common/auth/authentication.js (hunk 9 above)
Comment (suhaskatkuri-wq) [reply ui/app/common/auth/authentication.js:262]: Added an isAttemptingLogout flag - set to true on entry, reset once the hasDraftsForProvider promise resolves and on the direct logout path. So a double-click before the call resolves just returns early instead of firing attemptLogout twice.

File: ui/app/common/auth/authentication.js (hunk 10 above)
Comment (suhaskatkuri-wq) [reply ui/app/common/auth/authentication.js]: Removed scope.$apply() from the click handler. attemptLogout goes through $http/$q internally so digest already gets triggered, click and keyboard paths behave the same now.

File: ui/test/unit/common/auth/logOutDirective.spec.js (hunk 11 above)
Comment (suhaskatkuri-wq) [reply ui/test/unit/common/auth/logOutDirective.spec.js:217]: Added all 3 - keyboard shortcut with no drafts logs out directly, with drafts shows the warning, and with enableFormDraftFeature disabled logs out directly - to logOutDirective.spec.js.

File: ui/app/i18n/common/locale_en.json (hunk 12 above)
Comment (suhaskatkuri-wq) [reply ui/app/i18n/common/locale_en.json:6]: Added locale files for the other languages already supported in the app - ar, de, el, es, fr, hr, it, km, lo, pt, pt_BR, ru, vi, zh - under i18n/common/, translated the 4 drafts-warning keys in each. mergeLocaleFilesService already handles a missing file gracefully, so just the translations were missing, no code change needed there.

**Comment by vvkpd** [comment]
**`ui/app/registration/controllers/navigationController.js:17`** — the `$scope.logout` function here still calls `sessionService.destroy()` directly and bypasses `logoutService`:

```javascript
$scope.logout = function () {
    sessionService.destroy().then(function () {
        $window.location = "../home/index.html#/login";
    });
};
```

This is currently dead code — `registration/views/header.html` uses the `<a log-out>` directive which correctly goes through `logoutService`. But if this function ever gets wired up again, the draft warning will silently not show.

Either delete it, or update it to:
```javascript
$scope.logout = function () {
    logoutService.attemptLogout($scope);
};
```

**Review by vvkpd — APPROVED** [review]
